### DISCUSSION TITLE
Publishing logs from external system via API

### DISCUSSION THREAD

**Discussion by gbreavin** [jongpie/NebulaLogger#808] [Q&A] [answered]
I have a use case where I'm considering using Nebula logger to capture logs from an external system. There is a process where it would be useful for us to have log records from that external system to give more context to parts of the process that are in Salesforce (and also have logs).

My question is - is this a pattern that is supported or possible? Platform Events are the obvious choice, though need reconciling to make sure they are actually published; there's also Apex (would likely need to wrap in a web service, unless there's something already)... What about creating the Log and LogEntry records directly? Obviously possible, but concerned what we lose out on doing it direct vs an interface like the Apex methods. 

Are there any considerations (or other methods) that we should consider?

**Comment by jongpie** [comment, marked as answer]
Hi @gbreavin - great question! There is currently an open pull request - PR #767 - that will add a custom REST endpoint that will support saving external logs in Nebula Logger. The PR still needs a little more work before it's ready, but the idea is that external systems will be able to send their logging data to this endpoint in your Salesforce org, using [OpenTelemetry's (OTel) standard for logs](https://opentelemetry.io/docs/specs/otel/logs/data-model/), and Nebula Logger will convert it from OTel to Nebula Logger's data model. My hope is that by using the OTel standard, it will be easy for external systems to leverage - [OTel has SDKs for several popular languages](https://opentelemetry.io/docs/languages/), like Java, C#/.NET, Node/JS, python, Go, etc.

The other methods you mentioned (creating the platform events or creating the log/log entry records yourself) would also work, but you would have to build that yourself - there's nothing in Nebula Logger right now that would really help make that easier via the API.

**Reply by gbreavin** [reply]
Thanks @jongpie, the quick reply is much appreciated!

**Reply by jongpie** [reply]
No problem! My hope is to finish PR #767 in the next few weeks - I'll add another comment here once it's been released.

**Reply by JeroenSfdc** [reply]
We did just that, create a custom platform event, External_Log__e, allow create for our API-only user. Have simple event triggered flow, which calls invocable which calls Nebula using native methods. It's not great, but very helpful. In our case specifically to enable MuleSoft to log certain types of errors in their end into Salesforce with bespoke tag.

**Reply by coluccigiovanni16** [reply]
Hi @jongpie,  

Following @JeroenSfdc’s idea, it would be great to have a Platform Event as an entry point to save logs into Salesforce from external systems (like Mulesoft) with OTel format. What are your thoughts on this? Do you already have something in mind, or have you considered and ruled out this functionality?  
Additionally, another nice-to-have feature would be the ability to mass-insert logs. I’m not sure if this is already addressed in the PR discussed in this thread.  

Thanks!

**Reply by jonoddkol** [reply]
Hi @jongpie,

We're about to upgrade v4.16.0 right now and while that is happening I'd love to hear your thoughts on publishing LogEntryEvents from Mulesoft (or other API). 

We're starting out a project now where we aim to do this, but in lack of any documentation on what fields on the Platform Event object should be set we're feeling our way forward. Any input or suggested best practices with such a setup would be greatly appreciated! ☺️

**Reply by jongpie** [reply]
@coluccigiovanni16 apologies, I missed your previous comment from February! It's still a work in progress, but think that PR #767 (and subsequent enhancements) will provide all of that functionality:
- The PR itself will add the a custom REST endpoint to support sending data in OTel format
- Internally & by default, the REST endpoint will use the existing `LogEntryEvent__e` platform event
- It will also have an option to skip the platform events & instead write directly to the custom objects

@jonoddkol I think that the above info also applies for your use case as well - you definitely can insert `LogEntryEvent__e` platform events if you choose, but doing that directly isn't something that I'm planning to officially support. Instead, the recommendation will be to use OTel formatted JSON + the custom `LoggerRestResource` endpoint that's being added in PR #767.
